1: Classification modle with all data to determine buy or no buy. Balance the data, reduring the people that do not buy to match a similar value to that that have bought. (This will be the model that determines for the decision making of buy or no buy).

In [23]:
# Main Libraries
import pandas as pd  
import numpy as np
import statsmodels.api as sm
from statsmodels.formula.api import ols

In [2]:
#importing data wheel
file_path1 = 'nov24.csv'
wheel = pd.read_csv(file_path1)
#importing data wheel
file_path2 = 'nov17.csv'
wbyo = pd.read_csv(file_path2)

In [3]:
#adding the date each promotion was run.
wheel['date'] = pd.to_datetime('2023-11-24')
wbyo['date'] = pd.to_datetime('2023-11-17')

# Specify the columns to check for the condition on the wheel
columns_to_check1 = ['click_noch_primary', 'click_dreh', 'click_jetzt_spielen', 'click_gluecksrad_drehe', 'click_code_anzeigen']

# Use np.where to create the new column based on the condition on the wheel
wheel['gluecksrad_engagement'] = np.where(wheel[columns_to_check1].gt(0).any(axis=1), 1, 0)

# Specify the columns to check for the condition on the wbyo
columns_to_check2 = ['click_noch_primary', 'click_jetzt_einkaufen', 'click_gewinne', 'click_link_noch']

# Use np.where to create the new column based on the condition on the wbyo
wbyo['wbyo_engagement'] = np.where(wbyo[columns_to_check2].gt(0).any(axis=1), 1, 0)

#adding 0 to each dataframe for the new engagement columns so that the 2 dataframes can be merged.
wheel['wbyo_engagement'] = 0
wbyo['gluecksrad_engagement'] = 0

# Concatenate the DataFrames by adding the rows of 'wheel' to 'wbyo'
df = pd.concat([wheel, wbyo], ignore_index=True)

# Specify the columns to check for the condition on the wbyo
columns_to_check3 = ['wbyo_engagement', 'gluecksrad_engagement']

# Use np.where to create the new column based on the condition on the wbyo
df['gamification_engagement'] = np.where(df[columns_to_check3].gt(0).any(axis=1), 1, 0)

df

user_pseudo_id      country        city device_category  \
0         1.653411e+09      Germany      Berlin          mobile   
1         1.147704e+09      Germany  Mindelheim          mobile   
2         1.582437e+09  Switzerland       Bulle          mobile   
3         1.800091e+09      Germany   Frankfurt          mobile   
4         3.656791e+08      Germany         NaN          mobile   
...                ...          ...         ...             ...   
195586    2.082985e+09      Germany    Cuxhaven          mobile   
195587    2.129320e+09      Germany   Bielefeld          mobile   
195588    2.130216e+09      Germany  Hildesheim          mobile   
195589    2.131727e+09      Germany      Siegen          mobile   
195590    2.137817e+09      Germany     Wolgast          mobile   

       operating_system sourcemedium  new_user       first_event  \
0                   iOS          TBD         1  1700842391176249   
1                   iOS          TBD         1  1700848653518514   
2                   iOS          TBD         1  1700852395847646   
3                   iOS          TBD         1  1700828032485232   
4                   iOS          TBD         1  1700823107022159   
...                 ...          ...       ...               ...   
195586          Android          TBD         1  1700220385242808   
195587              iOS          TBD         1  1700222757561626   
195588              iOS          TBD         1  1700220383090692   
195589          Android          TBD         1  1700220929845757   
195590              iOS          TBD         0  1700222586111540   

              last_event  event_count  ...  click_noch_primary  \
0       1700842456749569           16  ...                   0   
1       1700848692399166           19  ...                   0   
2       1700852628033297           16  ...                   0   
3       1700828308513799           65  ...                   0   
4       1700823183682287            9  ...                   0   
...                  ...          ...  ...                 ...   
195586  1700220398841543            6  ...                   0   
195587  1700222798355484           13  ...                   0   
195588  1700220414150397           12  ...                   0   
195589  1700220929845757            3  ...                   0   
195590  1700222626280157            7  ...                   0   

        click_link_noch  click_dreh  click_jetzt_spielen  \
0                     0           0                    0   
1                     0           0                    0   
2                     0           0                    0   
3                     0           0                    0   
4                     0           0                    0   
...                 ...         ...                  ...   
195586                0           0                    0   
195587                0           0                    0   
195588                0           0                    0   
195589                0           0                    0   
195590                0           0                    0   

        click_gluecksrad_drehe click_code_anzeigen       date  \
0                            0                   0 2023-11-24   
1                            0                   0 2023-11-24   
2                            0                   0 2023-11-24   
3                            0                   0 2023-11-24   
4                            0                   0 2023-11-24   
...                        ...                 ...        ...   
195586                       0                   0 2023-11-17   
195587                       0                   0 2023-11-17   
195588                       0                   0 2023-11-17   
195589                       0                   0 2023-11-17   
195590                       0                   0 2023-11-17   

       gluecksrad_engagement  wbyo_engagement  gamification_engagement  
0                 

In [4]:
# Count the number of rows where 'wbyo_engagement' is 1
num_rows_value_1_wbyo_df = (df['wbyo_engagement'] == 1).sum()

print(f"Number of rows where 'wbyo_engagement' is 1 in df: {num_rows_value_1_wbyo_df}")

# Count the number of rows where 'gluecksrad_engagement' is 1
num_rows_value_1_wheel_df = (df['gluecksrad_engagement'] == 1).sum()

print(f"Number of rows where 'gluecksrad_engagement' is 1 in df: {num_rows_value_1_wheel_df}")

# Count the number of rows where 'gamification_engagement' is 1
num_rows_value_1_gamification_df = (df['gamification_engagement'] == 1).sum()

print(f"Number of rows where 'gamification_engagement' is 1 in df: {num_rows_value_1_gamification_df}")

Number of rows where 'wbyo_engagement' is 1 in df: 1826
Number of rows where 'gluecksrad_engagement' is 1 in df: 22056
Number of rows where 'gamification_engagement' is 1 in df: 23882


In [5]:
#rounding off decimal place to 2 decimals for code like "describe", ect
pd.set_option('display.float_format', lambda x: '%.2f' % x,)

In [6]:
# Convert 'first_event' and 'last_event' from Unix timestamps in microseconds to datetime
df['first_event'] = pd.to_datetime(df['first_event'], unit='us')
df['last_event'] = pd.to_datetime(df['last_event'], unit='us')

# Subtract 'first_event' from 'last_event' to get the time difference
df['time_diff'] = (df['last_event'] - df['first_event'])

# Convert the time difference to total seconds, creating a numerical column
df['time_spent_seconds'] = df['time_diff'].dt.total_seconds()

df

user_pseudo_id      country        city device_category  \
0        1653410570.17      Germany      Berlin          mobile   
1        1147704469.17      Germany  Mindelheim          mobile   
2        1582437117.17  Switzerland       Bulle          mobile   
3        1800090715.17      Germany   Frankfurt          mobile   
4         365679149.17      Germany         NaN          mobile   
...                ...          ...         ...             ...   
195586   2082984818.17      Germany    Cuxhaven          mobile   
195587   2129320152.17      Germany   Bielefeld          mobile   
195588   2130215726.17      Germany  Hildesheim          mobile   
195589   2131727157.17      Germany      Siegen          mobile   
195590   2137816702.17      Germany     Wolgast          mobile   

       operating_system sourcemedium  new_user                first_event  \
0                   iOS          TBD         1 2023-11-24 16:13:11.176249   
1                   iOS          TBD         1 2023-11-24 17:57:33.518514   
2                   iOS          TBD         1 2023-11-24 18:59:55.847646   
3                   iOS          TBD         1 2023-11-24 12:13:52.485232   
4                   iOS          TBD         1 2023-11-24 10:51:47.022159   
...                 ...          ...       ...                        ...   
195586          Android          TBD         1 2023-11-17 11:26:25.242808   
195587              iOS          TBD         1 2023-11-17 12:05:57.561626   
195588              iOS          TBD         1 2023-11-17 11:26:23.090692   
195589          Android          TBD         1 2023-11-17 11:35:29.845757   
195590              iOS          TBD         0 2023-11-17 12:03:06.111540   

                       last_event  event_count  ...  click_dreh  \
0      2023-11-24 16:14:16.749569           16  ...           0   
1      2023-11-24 17:58:12.399166           19  ...           0   
2      2023-11-24 19:03:48.033297           16  ...           0   
3      2023-11-24 12:18:28.513799           65  ...           0   
4      2023-11-24 10:53:03.682287            9  ...           0   
...                           ...          ...  ...         ...   
195586 2023-11-17 11:26:38.841543            6  ...           0   
195587 2023-11-17 12:06:38.355484           13  ...           0   
195588 2023-11-17 11:26:54.150397           12  ...           0   
195589 2023-11-17 11:35:29.845757            3  ...           0   
195590 2023-11-17 12:03:46.280157            7  ...           0   

        click_jetzt_spielen  click_gluecksrad_drehe  click_code_anzeigen  \
0                         0                       0                    0   
1                         0                       0                    0   
2                         0                       0                    0   
3                         0                       0                    0   
4                         0                       0                    0   
...                     ...                     ...                  ...   
195586                    0                       0                    0   
195587                    0                       0                    0   
195588                    0                       0                    0   
195589                    0                       0                    0   
195590                    0                       0                    0   

             date gluecksrad_engagement  wbyo_engagement  \
0      2023-11-24                     0                0   
1      2023-11-24                     0                0   
2      2023-11-24                     0                0   
3      2023-11-24                     0                0   
4      2023-11-24                     0                0   
...           ...                   ...              ...   
195586 2023-11-17                     0                0   
195587 2023-11-17                     0                0   
195588 2023-11-17   

In [7]:
# List of specific columns to drop
columns_to_drop = ['time_diff', 'coupons', 'quantity', 'sourcemedium', 'first_event', 'last_event', 'click_jetzt_einkaufen', 'click_gewinne', 'click_primary_header_wrapper', 'click_noch_primary', 'click_link_noch', 'click_dreh', 'click_jetzt_spielen', 'click_gluecksrad_drehe', 'click_code_anzeigen']

# Dropping the columns from the DataFrame
df = df.drop(columns=columns_to_drop, errors='ignore')  # errors='ignore' to avoid errors if a column is missing

# Display the DataFrame to verify the columns have been dropped
df.info(verbose=True)

df1 = df

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195591 entries, 0 to 195590
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   user_pseudo_id           195591 non-null  float64       
 1   country                  195588 non-null  object        
 2   city                     176991 non-null  object        
 3   device_category          195591 non-null  object        
 4   operating_system         195590 non-null  object        
 5   new_user                 195591 non-null  int64         
 6   event_count              195591 non-null  int64         
 7   items_viewed             195591 non-null  int64         
 8   add_to_carts             195591 non-null  int64         
 9   checkouts                195591 non-null  int64         
 10  purchases                195591 non-null  int64         
 11  total_revenue            195591 non-null  float64       
 12  sessions        

In [8]:
# Create dummy variables for 'device_category'
device_category_dummies = pd.get_dummies(df['device_category'], prefix='device_category')

# Drop the original 'device_category' column from 'df'
df = df.drop('device_category', axis=1)

# Concatenate the dummy variables DataFrame with the original 'df' DataFrame
df = pd.concat([df, device_category_dummies], axis=1)

In [9]:
# Filter 'df' to keep only rows where the 'country' column is 'Germany'
#This is becuase purchases can only be made in germany for this specific store.
df = df[df['country'] == 'Germany']

# Create dummy variables for 'operating_system'
operating_system_dummies = pd.get_dummies(df['operating_system'], prefix='operating_system')

# Drop the original 'operating_system' column from 'df'
df = df.drop('operating_system', axis=1)

# Concatenate the dummy variables DataFrame with the original 'df' DataFrame
df = pd.concat([df, operating_system_dummies], axis=1)

#Dropping the Country and City due to country only being germany and city scope being to wide
#user_pseudo_id is removed as it is not needed either for the OLS

# List of specific columns to drop
columns_to_drop = ['country', 'city', 'user_pseudo_id']

# Dropping the columns from the DataFrame
df = df.drop(columns=columns_to_drop, errors='ignore')  # errors='ignore' to avoid errors if a column is missing

# Convert 'date' column to dummy variable: 0 for '2023-11-17' and 1 for '2023-11-24'
df['date'] = df['date'].apply(lambda x: 1 if x == pd.Timestamp('2023-11-24') else 0)

# Display the DataFrame to verify the columns have been dropped
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
Int64Index: 187930 entries, 0 to 195590
Data columns (total 23 columns):
 #   Column                       Non-Null Count   Dtype  
---  ------                       --------------   -----  
 0   new_user                     187930 non-null  int64  
 1   event_count                  187930 non-null  int64  
 2   items_viewed                 187930 non-null  int64  
 3   add_to_carts                 187930 non-null  int64  
 4   checkouts                    187930 non-null  int64  
 5   purchases                    187930 non-null  int64  
 6   total_revenue                187930 non-null  float64
 7   sessions                     187930 non-null  int64  
 8   date                         187930 non-null  int64  
 9   gluecksrad_engagement        187930 non-null  int64  
 10  wbyo_engagement              187930 non-null  int64  
 11  gamification_engagement      187930 non-null  int64  
 12  time_spent_seconds           187930 non-null  float64
 13 

In [10]:
df2 = df

In [11]:
#Using DF1 as dataframe for ANOVA as it contains objects and dummy variables.
cdf = df1


In [12]:
# Display the DataFrame to verify the columns have been dropped
df1.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 195591 entries, 0 to 195590
Data columns (total 18 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   user_pseudo_id           195591 non-null  float64       
 1   country                  195588 non-null  object        
 2   city                     176991 non-null  object        
 3   device_category          195591 non-null  object        
 4   operating_system         195590 non-null  object        
 5   new_user                 195591 non-null  int64         
 6   event_count              195591 non-null  int64         
 7   items_viewed             195591 non-null  int64         
 8   add_to_carts             195591 non-null  int64         
 9   checkouts                195591 non-null  int64         
 10  purchases                195591 non-null  int64         
 11  total_revenue            195591 non-null  float64       
 12  sessions        

## Running the ANOVA test

#### 1. ANOVA Test for Purchase Likelihood
Null Hypothesis (H0): The null hypothesis for the purchase likelihood test states that there is no difference in the likelihood of making a purchase among the different levels of gamification_engagement. In other words, the engagement levels do not affect the likelihood of a purchase.
- H0: μ1 = μ2 = ... = μk (where μ represents the mean purchase likelihood for each level of gamification_engagement)

Alternative Hypothesis (H1): The alternative hypothesis states that at least one level of gamification_engagement differs significantly in terms of purchase likelihood. This implies that how much users engage with gamification features affects their likelihood to make a purchase.
- H1: Not all μ are equal. (At least one group mean is different.)

In [25]:
#Running ANOVA for Purchase Likelihood

# Convert purchases to a binary outcome if necessary
df1['purchases'] = df1['purchases'].apply(lambda x: 1 if x > 0 else 0)

# Fit the model
model_purchases = ols('purchases ~ C(gamification_engagement)', data=df1).fit()

# Perform ANOVA
anova_purchases = sm.stats.anova_lm(model_purchases, typ=2)
print("ANOVA results for Purchase Likelihood:")
print(anova_purchases)

Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
ANOVA results for Purchase Likelihood:
                             sum_sq        df       F  PR(>F)
C(gamification_engagement)   111.45      1.00 2011.53    0.00
Residual                   10836.71 195589.00     NaN     NaN


#### ANOVA Results for Purchase Likelihood
1. Sum of Squares (SS)
    - Between Groups (C(gamification_engagement)): The sum of squares due to the gamification_engagement is 111.45. This value represents the variation attributable to differences between the different levels of gamification engagement.
    - Residual: The residual sum of squares is 10836.71, representing the variation within groups or unexplained by the model.
2. Degrees of Freedom (df)
    - Between Groups: The degrees of freedom for the gamification engagement group is 1, indicating two levels of engagement (usually 0 and 1 in binary cases).
    - Residual: The degrees of freedom for the residuals is 195,589, which is the number of observations minus the number of groups.
3. F-Statistic: The F-statistic is 2011.53, which is quite large. It is calculated as the mean square of the group (MS between) divided by the mean square of the error (MS within).
4. P-value: The p-value is 0.00, which indicates that the differences in purchase likelihood between the two levels of gamification engagement are statistically significant. This p-value effectively rejects the null hypothesis that there are no differences in mean purchase likelihood across the groups defined by gamification engagement.

#### 2. ANOVA Test for Total Revenue
Null Hypothesis (H0): The null hypothesis for the total revenue test states that there is no difference in the average revenue generated among the different levels of gamification_engagement. Essentially, whether or not individuals engage in gamification does not influence the amount of revenue they generate.
- H0: μ1 = μ2 = ... = μk (where μ represents the mean total revenue for each level of gamification_engagement)

Alternative Hypothesis (H1): The alternative hypothesis contends that at least one level of gamification_engagement leads to a significantly different average revenue. This suggests that engagement levels impact the revenue amounts.
- H1: Not all μ are equal. (At least one group mean is different.)

In [26]:
#Running ANOVA for Total Revenue

# Fit the model
model_revenue = ols('total_revenue ~ C(gamification_engagement)', data=df1).fit()

# Perform ANOVA
anova_revenue = sm.stats.anova_lm(model_revenue, typ=2)
print("ANOVA results for Total Revenue:")
print(anova_revenue)

Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
ANOVA results for Total Revenue:
                                  sum_sq        df       F  PR(>F)
C(gamification_engagement)   26364139.32      1.00 1656.51    0.00
Residual                   3112897402.57 195589.00     NaN     NaN


#### ANOVA Results for Total Revenue
1. Sum of Squares (SS)
    - Between Groups: The sum of squares due to gamification_engagement for total revenue is about 26,364,139.32, indicating significant variation in revenue attributable to differences in gamification engagement.
    - Residual: The residual sum of squares is considerably larger at 3,112,897,402.57, suggesting a lot of variation within groups that the model does not explain.
2. Degrees of Freedom (df)
    - As with the purchase likelihood, there’s 1 degree of freedom for gamification engagement and 195,589 for residuals.
3. F-Statistic: The F-statistic for revenue is 1656.51. This is also a substantial value and suggests that the means between groups are significantly different.
4. P-value: The p-value is again 0.00, indicating strong statistical significance. This suggests that the gamification engagement significantly affects the revenue, with different levels of engagement leading to different average revenues.

#### Interpretation and Implications
- Statistical Significance: Both models show that gamification engagement has a statistically significant effect on both the likelihood of making a purchase and the amount of revenue generated. The extremely low p-values in both cases allow us to reject the null hypothesis for both models.
- Practical Significance: Given the large F-statistics, the effect size (i.e., the difference between group means) is likely substantial in both cases. However, it's crucial to consider the practical implications of these findings — specifically, how significant the differences are in practical, monetary terms for revenue, and conversion rate terms for purchases.
- Further Actions: Given these results, you might consider strategies to increase gamification engagement as a way to boost both purchase likelihood and revenue. Additionally, further analysis could explore which aspects of gamification are most effective, or how these effects hold across different segments of users.

Overall, these results are very promising for justifying investments in gamification strategies as part of marketing and customer engagement efforts.

### General understanding & Interpretation of Results
- F-statistic and p-value: In the ANOVA table, check the F-statistic and corresponding p-value. A significant p-value (typically <.05) suggests that there are significant differences in the means across the groups defined by gamification_engagement.
- Sum of Squares (SS): This part of the output tells you how much variance is explained by the groups.
- Degrees of Freedom (df): This indicates the number of levels in the categorical variable minus one.
- Mean Square (MS): This is the SS divided by the df and is used in calculating the F-statistic.